From c5c314519f997b43f95df05c0cc8ea4269406a40 Mon Sep 17 00:00:00 2001
From: Ryan Bruno <ryan@rbruno.com>
Date: Wed, 24 Jun 2020 21:55:57 -0400
Subject: [PATCH] Read timeouts.

---
 coms/smtpd.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/coms/smtpd.c b/coms/smtpd.c
index 4dff8d5..ddc5a63 100644
--- a/coms/smtpd.c
+++ b/coms/smtpd.c
@@ -65,6 +65,17 @@ char* read_some(int fd, char* b, size_t s)
 
     for (s--;;) {
 
+        fd_set fds; FD_ZERO(&fds); FD_SET(fd, &fds);
+
+        /* 5 Second timeout */
+        struct timeval timeout;
+        timeout.tv_sec = 5;
+        timeout.tv_usec = 0;
+
+        if (select(fd + 1, &fds, NULL, NULL, &timeout) < 1)
+            /* Timeout or error */
+            break;
+
         if ((r = read(fd, b, s)) == 0)
             /* Either EOF or socket closed */
             break;
-- 
2.27.0

